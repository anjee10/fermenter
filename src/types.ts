export interface IGherkinMethods {
  Scenario: (name: string) => IScenarioFluid;
  ScenarioOutline: (name: string) => IScenarioFluid;
  /**
   * Occasionally you'll find yourself repeating the same Given steps in all of the scenarios in a feature.
   * Since it is repeated in every scenario, this is an indication that those steps are not essential to describe the scenarios; they are incidental details. You can literally move such Given steps to the background, by grouping them under a Background section.
   *
   * - A Background allows you to add some context to the scenarios in the feature. It can contain one or more Given steps.
   * - A Background is run before each scenario, but after any Before hooks. In your feature file, put the Background before the first Scenario.
   * - You can only have one set of Background steps per feature. If you need different Background steps for different scenarios, you'll need to split them into different feature files.
   */
  Background: (fn: any) => IGivenFluid; // FIXME:
  Hooks?: {}; // FIXME:
}

// TODO: add But as And alias if necessary

export interface IAndFluid {
  And: IFluidFn<IAndFluid>;
}

export interface IScenarioFluid {
  Given: IFluidFn<IGivenFluid>;
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid>;
}

export interface IBackgroundFluid {
  Given: IFluidFn<IAndFluid>;
}

export interface IScenarioFluid {
  Given: IFluidFn<IGivenFluid>;
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid>;
}

export interface IScenarioOutlineExamples<N = any> {
  Examples: IFluidFn<N>;
}

export interface IScenarioOutlineFluid extends IScenarioOutlineExamples<IScenarioOutlineFluid> {
  Given: IFluidFn<IGivenFluid>;
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid & IScenarioOutlineExamples<{}>>;
}

export interface IGivenFluid {
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid>;
  And: IFluidFn<IGivenFluid>;
}

export interface IWhenFluid {
  Then: IFluidFn<IAndFluid>;
  And: IFluidFn<IAndFluid>;
}

export type IFluidFn<Fluid> = (key: RegExp|string, fn: IFluidFnCallback) => Fluid;
export type IFluidFnCallback = (state: any, ...params: any[]) => void;
