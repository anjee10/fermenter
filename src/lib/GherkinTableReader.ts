import { IGherkinAstTableRow, IGherkinTableParam } from '../types';

/**
 * A gherkin feature file's rows:
 *
 *   | a | b |
 *   | 1 | 2 |
 *   | 3 | 4 |
 *
 * Becomes `IGherkinAstTableRows[]`
 *
 * @example
 *
 * const table = GherkinTableReader({ rows })
 */
export function GherkinTableReader ({ rows: inputRows = [] }: {
  rows: IGherkinAstTableRow[],
}): IGherkinTableParam {
  const rowValues = inputRows
    .map(({ cells }) => cells.map(({ value }) => value));

  const getLeft = () => rowValues.map((cells) => cells[0]);
  const getTop = () => rowValues[0];

  const dict = <IGherkinTableParam['dict']> (() => {
    const keys = headers();

    const contentRows = rowValues.slice(1);

    const tuples = keys.map((key, index): [string, string[]] => {
      return [
        key,
        contentRows.map((cells) => cells[index]),
      ];
    });

    return new Map(tuples);
  });

  dict.left = () => {
    const keys = rowValues.map((cells) => cells[0]);

    const mapTuples = rowValues.map((cells, index): [string, string[]] => {
      /** The cells, after the left-most */
      const contentCells = cells.slice(1);

      return [keys[index], contentCells];
    });

    return new Map(mapTuples);
  };

  dict.matrix = () => {
    const topKeys = getTop()
      .slice(1); // Skip blank

    const leftKeys = getLeft()
      .slice(1); // Skip blank

    const contentRows = rowValues
      .slice(1) // Skip top
      .map((cells) => cells.slice(1)); // Skip left

    const leftTuples = leftKeys.map((leftKey, leftIndex): [string, Map<string, string>] => {
      const topTuples = topKeys.map((topKey, topIndex): [string, string] => {
        const cells = contentRows[topIndex];

        return [topKey, cells[leftIndex]];
      });

      return [leftKey, new Map(topTuples)];
    });

    return new Map(leftTuples);
  };

  const rows = <IGherkinTableParam['rows']> (() => {
    return inputRows.map(({ cells }) => cells.map(({ value }) => value));
  });

  rows.mapByTop = () => {
    const keys = getTop();

    const contentRows = rowValues.slice(1);

    return contentRows.map((cells) => {
      const tuples = keys.map((key, index): [string, string] => {
        return [key, cells[index]];
      });

      return new Map(tuples);
    });
  };

  const headers = <IGherkinTableParam['headers']> (() => getTop());

  return { dict, headers, rows };
}
