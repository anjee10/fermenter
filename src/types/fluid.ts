import { IMatch } from '../GherkinEngine';

export interface IGherkinMethods {
  Scenario: (match: IMatch) => IScenarioFluid;
  ScenarioOutline: (match: IMatch) => IScenarioOutlineFluid;

  /**
   * Occasionally you'll find yourself repeating the same Given steps in all of the scenarios in a feature.
   * Since it is repeated in every scenario, this is an indication that those steps are not essential to describe the scenarios; they are incidental details. You can literally move such Given steps to the background, by grouping them under a Background section.
   *
   * - A Background allows you to add some context to the scenarios in the feature. It can contain one or more Given steps.
   * - A Background is run before each scenario, but after any Before hooks. In your feature file, put the Background before the first Scenario.
   * - You can only have one set of Background steps per feature. If you need different Background steps for different scenarios, you'll need to split them into different feature files.
   */
  Background: (match?: IMatch) => IBackgroundFluid;
  Hooks?: {}; // FIXME:
}

//
// FLUID INTERFACES
//

export interface IAndFluid {
  And: IFluidFn<IAndFluid>;
}

export interface IScenarioFluid {
  Given: IFluidFn<IGivenFluid>;
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid>;
}

export interface IBackgroundFluid {
  Given: IFluidFn<IAndFluid>;
}

export interface IScenarioFluid {
  Given: IFluidFn<IGivenFluid>;
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid>;
}

export interface IScenarioOutlineExamplesFluid<N = any> {
  Examples: IFluidFn<N>;
}

export interface IScenarioOutlineFluid extends IScenarioOutlineExamplesFluid<IScenarioOutlineFluid> {
  Given: IFluidFn<IGivenFluid>;
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid & IScenarioOutlineExamplesFluid<IScenarioOutlineFluid>>;
}

export interface IGivenFluid {
  When: IFluidFn<IWhenFluid>;
  Then: IFluidFn<IAndFluid>;
  And: IFluidFn<IGivenFluid>;
}

export interface IWhenFluid {
  Then: IFluidFn<IAndFluid>;
  And: IFluidFn<IAndFluid>;
}

export type IFluidFn<Fluid> = (key: RegExp|string, fn: IFluidFnCallback) => Fluid;
export type IFluidFnCallback<State = any> = (state: State, ...params: any[]) => State;
