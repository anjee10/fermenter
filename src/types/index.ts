
// tslint:disable-next-line
export type Omit<T, K extends keyof T> = T extends any ? Pick<T, Exclude<keyof T, K>> : never;
export type UnwrapPromise<T> = T extends Promise<infer U> ? U : T;
export type AsyncReturnType<T extends (...args: any[]) => any> = UnwrapPromise<ReturnType<T>>;

export * from './fluid';
export * from './ast';
export * from './parser';
