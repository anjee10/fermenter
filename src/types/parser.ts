import {
  IBackgroundFluid, IFluidFnCallback, IGherkinAstScenario, IGherkinAstScenarioOutline,
  IScenarioFluid, IScenarioOutlineFluid,
} from '.';
import { IGherkinAstBackground, IGherkinAstEntity, IGherkinAstExamples, IGherkinAstFeature, IGherkinAstStep, IGherkinAstTableRow } from './ast';

// tslint:disable-next-line

export type IMatch = string | RegExp;

export type IGherkinCollectionItemShape = IGherkinAstEntity | IGherkinAstStep;
export type IGherkinCollectionItemIndex = IGherkinAstEntity & IGherkinAstStep;

export type IGherkinParams = string | number | IGherkinAstTableRow;

export interface IGherkinTableParam {

  /**
   * With Gherkin:
   * ```
   * | a | b |
   * | 1 | 2 |
   * ```
   * @example
   *
   * [ ...table.rows() ]
   *
   * // returns
   *
   * [
   *   [ 'a', 'b' ],
   *   [ '1', '2' ],
   * ]
   *
   */
  rows: {
    (): string[][];

    /**
     * With Gherkin:
     *
     * ```
     * | a | b |
     * | 1 | 2 |
     * | 3 | 4 |
     * ```
     *
     * @example
     *
     * table.rows.mapByTop()
     *
     * // returns an array of Map's, mapped to the header as keys
     *
     * [
     *   [ ['a', '1'], ['b', '2'] ],
     *   [ ['a', '3'], ['b', '4'] ],
     * ]
     */
    mapByTop (): Array<Map<string, string>>;
  };

  /**
   * With Gherkin:
   * ```
   * | a | b |
   * | 1 | 2 |
   * | 1 | 2 |
   * ```
   * @example
   *
   * [ ...table.dict() ]
   *
   * // returns a Map, mapped to the header as keys
   *
   * [
   *   [ 'a', ['1', '1'] ],
   *   [ 'b', ['2', '2'] ],
   * ]
   *
   */
  dict: {
    (): Map<string, string[]>;

    /**
     * With Gherkin:
     * ```
     * | a | 1 | 3 |
     * | b | 2 | 4 |
     * ```
     * @example
     *
     * [ ...table.dict.left() ]
     *
     * // returns a Map, mapped to the left column as keys
     *
     * [
     *   [ 'a', ['1', '3'] ],
     *   [ 'b', ['2', '4'] ],
     * ]
     *
     */
    left (): Map<string, string[]>;

    /**
     * With Gherkin:
     * ```
     * |   | x | y |
     * | a | 1 | 3 |
     * | b | 2 | 4 |
     * ```
     * @example
     *
     * [ ...table.dict.matrix() ]
     *
     * // returns a Map, mapped to the left column as keys into a Map of head column keys
     *
     * [
     *   [ 'a', [[ 'x', '1' ], ['y', '3']] ],
     *   [ 'b', [[ 'x', '2' ], ['y', '4']] ],
     * ]
     *
     */
    matrix (): Map<string, Map<string, string>>
  };

  /**
   * With Gherkin:
   * ```
   * | a | b |
   * | 1 | 2 |
   * | 1 | 2 |
   * ```
   * @example
   *
   * [ ...table.headers() ]
   *
   * // returns
   *
   * ['a', 'b']
   */
  headers (): string[];
}

export interface IGherkinFeatureMappings<Ge = any> {
  match: IMatch;
  gherkin: Ge;
}

export interface IScenarioBuilder<
  S extends IGherkinScenario | IGherkinScenarioOutline = IGherkinScenario
> {
  steps: IScenarioFluid;
  scenario: S;
}

export interface IBackgroundBuilder {
  steps: IBackgroundFluid;
  background: IGherkinBackground;
}

/**
 * Produces a subset ScenarioOutline that looks like this:
 * @example { Given, When, Then, Examples: { operations } }
 */
export interface IScenarioOutlineBuilder {
  steps: IScenarioOutlineFluid;
  scenarioOutline: IGherkinScenarioOutline;
}

export type IGherkinOperationStore<
  G extends IGherkinCollectionItemShape = IGherkinCollectionItemShape
> = Map<IMatch, {
  fn: IFluidFnCallback,
  name: string;
  gherkin: G;
  params: any[], // TODO: this needs to be array of string/int/data table
}>;

export interface IGherkinScenarioBase {
  match: IMatch;
  name: IGherkinAstScenario['name'];
  Given?: IGherkinOperationStore<IGherkinAstStep>;
  When?: IGherkinOperationStore<IGherkinAstStep>;
  Then?: IGherkinOperationStore<IGherkinAstStep>;
}

export interface IGherkinScenario extends IGherkinScenarioBase {
  gherkin: IGherkinAstScenario;
}

export interface IGherkinScenarioOutline extends IGherkinScenarioBase {
  gherkin: IGherkinAstScenarioOutline;
  Examples: IGherkinOperationStore<IGherkinAstExamples>;
}

export interface IGherkinBackground {
  match: IMatch;
  name: IGherkinAstBackground['name'];
  gherkin: IGherkinAstBackground;
  Given: IGherkinOperationStore<IGherkinAstStep>;
}

export interface IGherkinFeature {
  gherkin: IGherkinAstFeature;
  name: IGherkinAstFeature['name'];

  Background?: IGherkinBackground;
  Scenarios: Map<IMatch, IGherkinScenario>;
  ScenarioOutlines: Map<IMatch, IGherkinScenarioOutline>;
}
