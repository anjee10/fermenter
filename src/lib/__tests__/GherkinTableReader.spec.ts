import { IGherkinAstTableCell, IGherkinAstTableRow } from '../../types';
import { GherkinTableReader } from '../GherkinTableReader';

function structureAsGherkinAstRows (rows: string[][]): IGherkinAstTableRow[] {
  return rows.map((cells): IGherkinAstTableRow => {
    return {
      type: 'TableRow',
      location: {} as any,
      cells: cells.map((value): IGherkinAstTableCell => {
        return {
          location: {} as any,
          type: 'TableCell',
          value,
        };
      }),
    };
  });
}

function rowsToGherkinTableLines (rows: string[][]): string[] {
  return rows.map((cells) => `| ${cells.join(' | ')} |`);
}

describe('GherkinTableReader', () => {

  const simpleContentTable = [
    ['1', '2'],
    ['3', '4'],
  ];

  const topHeaderTable = [
    ['a', 'b'],
    ['1', '3'],
    ['2', '4'],
  ];

  const leftHeaderTable = [
    ['a', '1', '2'],
    ['b', '3', '4'],
  ];

  const matrixTable = [
    [' ', 'x', 'y'],
    ['a', '1', '2'],
    ['b', '3', '4'],
  ];

  const matrixTableAsymmetric = [
    [' ', 'x', 'y'],
    ['a', '1', '2'],
    ['b', '3', '4'],
    ['c', '5', '6'],
    ['d', '7', '8'],
  ];

  it([
    `table.rows() from a headerless table:`,
    ...rowsToGherkinTableLines(simpleContentTable),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(simpleContentTable) });

    expect(table.rows()).toMatchSnapshot();
  });

  it([
    `table.rows.mapByTop() from a table with a header:`,
    ...rowsToGherkinTableLines(topHeaderTable),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(topHeaderTable) });

    expect(table.rows.mapByTop()).toMatchSnapshot();
  });

  it([
    `table.dict() from a table with a header:`,
    ...rowsToGherkinTableLines(topHeaderTable),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(topHeaderTable) });

    expect(table.dict()).toMatchSnapshot();
  });

  it([
    `table.dict() from a headerless table:`,
    ...rowsToGherkinTableLines(simpleContentTable),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(simpleContentTable) });

    expect(table.dict()).toMatchSnapshot();
  });

  it([
    `table.dict.left() from a table with a header on the left:`,
    ...rowsToGherkinTableLines(leftHeaderTable),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(leftHeaderTable) });

    expect(table.dict.left()).toMatchSnapshot();
  });

  it([
    `table.dict.matrix() from a table with a header on the left and the top:`,
    ...rowsToGherkinTableLines(matrixTable),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(matrixTable) });

    expect(table.dict.matrix()).toMatchSnapshot();
  });

  it([
    `table.dict.matrix() with asymmetric top and left keys:`,
    ...rowsToGherkinTableLines(matrixTableAsymmetric),
    ``,
  ].join('\n'), () => {
    const table = GherkinTableReader({ rows: structureAsGherkinAstRows(matrixTableAsymmetric) });

    expect(table.dict.matrix()).toMatchSnapshot();
  });
});
