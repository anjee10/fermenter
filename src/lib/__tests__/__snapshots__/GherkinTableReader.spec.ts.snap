// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GherkinTableReader table.dict() from a headerless table:
| 1 | 2 |
| 3 | 4 |
 1`] = `
Object {
  "1": Array [
    "3",
  ],
  "2": Array [
    "4",
  ],
}
`;

exports[`GherkinTableReader table.dict() from a table with a header:
| a | b |
| 1 | 3 |
| 2 | 4 |
 1`] = `
Object {
  "a": Array [
    "1",
    "2",
  ],
  "b": Array [
    "3",
    "4",
  ],
}
`;

exports[`GherkinTableReader table.dict.left() from a table with a header on the left:
| a | 1 | 2 |
| b | 3 | 4 |
 1`] = `
Object {
  "a": Array [
    "1",
    "2",
  ],
  "b": Array [
    "3",
    "4",
  ],
}
`;

exports[`GherkinTableReader table.dict.matrix() from a table with a header on the left and the top:
|   | x | y |
| a | 1 | 2 |
| b | 3 | 4 |
 1`] = `
Object {
  "a": Object {
    "x": "1",
    "y": "3",
  },
  "b": Object {
    "x": "2",
    "y": "4",
  },
}
`;

exports[`GherkinTableReader table.rows() from a headerless table:
| 1 | 2 |
| 3 | 4 |
 1`] = `
Array [
  Array [
    "1",
    "2",
  ],
  Array [
    "3",
    "4",
  ],
]
`;

exports[`GherkinTableReader table.rows.mapByTop() from a table with a header:
| a | b |
| 1 | 3 |
| 2 | 4 |
 1`] = `
Array [
  Object {
    "a": "1",
    "b": "3",
  },
  Object {
    "a": "2",
    "b": "4",
  },
]
`;
